### PR TITLE
Allow an administrator to debug a node by launching a tools pod on that node

### PR DESCRIPTION
**Pull request by smarterclayton**
Runs as root without constraints on the node, intended for administrative
level debugging.

```
$ oc debug node/ip-10-0-166-213.ec2.internal
Debugging with pod/ip-10-0-166-213ec2internal-debug, original command: <image entrypoint>
Waiting for pod to start ...
If you don't see a command prompt, try pressing enter.
sh-4.2#
```

### CODE REVIEW

**Comment by smarterclayton** [comment]
I'm going to add `tools` to openshift namespace using samples operator @bparees.  I'm not going to split this across origin / OCP for now.

**Comment by chancez** [comment]
Should the pod be privileged?

**Comment by smarterclayton** [comment]
It has to be to be outside SELinux confinement and view host files.

**Comment by smarterclayton** [comment]
This is ready for a first pass review @soltysh (I expect we'll iterate based on feedback)

File: pkg/oc/cli/debug/debug.go [hunk 1]
```diff
@@ -345,11 +348,15 @@ func (o *DebugOptions) RunDebug() error {
 		ObjectMeta: template.ObjectMeta,
 		Spec:       template.Spec,
 	}
-	pod.Name, pod.Namespace = fmt.Sprintf("%s-debug", generateapp.MakeSimpleName(infos[0].Name)), infos[0].Namespace
+	ns := infos[0].Namespace
```
Comment (soltysh) [inline pkg/oc/cli/debug/debug.go]: Why is this needed, I'd expect every info object to have a namespace unless that's a non-namespaced one (like node) but in that case empty namespace is still fine

File: pkg/oc/cli/debug/debug.go [hunk 2]
```diff
@@ -725,6 +740,66 @@ func containerNames(pod *corev1.Pod) []string {
...
 	switch t := object.(type) {
+	case *corev1.Node:
+		if len(o.NodeName) > 0 {
+			return nil, fmt.Errorf("you may not set --node-name when debugging a node")
+		}
+		if o.AsNonRoot || o.AsUser > 0 {
+			// TODO: allow --as-root=false to skip all the namespaces except network
+			return nil, fmt.Errorf("can't debug nodes without running as the root user")
+		}
+		image := "registry.access.redhat.com/rhel7/rhel-tools"
```
Comment (soltysh) [inline pkg/oc/cli/debug/debug.go]: `image := o.Image` and then check that in the if below, otherwise you never allow specifying this.

File: pkg/oc/cli/debug/debug.go [hunk 3]
```diff
@@ -725,6 +740,66 @@ func containerNames(pod *corev1.Pod) []string {
...
+						},
+					},
+				},
+				RestartPolicy: corev1.RestartPolicyNever,
+				Containers: []corev1.Container{
+					{
+						Name:  "container-00",
+						Image: image,
+						Resources: corev1.ResourceRequirements{
+							Requests: corev1.ResourceList{
```
Comment (soltysh) [inline pkg/oc/cli/debug/debug.go]: Do we need to set these?

File: pkg/oc/cli/debug/debug.go [hunk 4]
```diff
@@ -725,6 +740,66 @@ func containerNames(pod *corev1.Pod) []string {
...
+	case *corev1.Node:
+		if len(o.NodeName) > 0 {
+			return nil, fmt.Errorf("you may not set --node-name when debugging a node")
+		}
+		if o.AsNonRoot || o.AsUser > 0 {
+			// TODO: allow --as-root=false to skip all the namespaces except network
+			return nil, fmt.Errorf("can't debug nodes without running as the root user")
+		}
+		image := "registry.access.redhat.com/rhel7/rhel-tools"
+		// container image is overwritten, but avoid
```
Comment (soltysh) [inline pkg/oc/cli/debug/debug.go]: not sure I understand what you mean by that comment?

File: pkg/oc/cli/debug/debug.go [hunk 5]
```diff
@@ -191,6 +193,7 @@ func NewCmdDebug(fullName string, f kcmdutil.Factory, streams genericclioptions.
 	cmd.Flags().StringVar(&o.NodeName, "node-name", o.NodeName, "Set a specific node to run on - by default the pod will run on any valid node")
 	cmd.Flags().BoolVar(&o.AsRoot, "as-root", o.AsRoot, "If true, try to run the container as the root user")
 	cmd.Flags().Int64Var(&o.AsUser, "as-user", o.AsUser, "Try to run the container as a specific user UID (note: admins may limit your ability to use this flag)")
+	cmd.Flags().StringVar(&o.Image, "image", o.Image, "Override the image used by the targeted container.")
```
Comment (soltysh) [inline pkg/oc/cli/debug/debug.go]: Maybe worth pointing this applies only to node debugging, and adding validation which ensure this is not set in other cases or at least a warning so we don't get issues where someone will complain that they set the image and it's doing nothing.

**Comment by deads2k** [comment]
I can't run `crictl` in this pod, so i can't look at failing static pod logs

File: pkg/oc/cli/debug/debug.go (hunk 1 above)
Comment (smarterclayton) [reply pkg/oc/cli/debug/debug.go]: It's not, because we have to create a pod with a namespace, or the create call will fail.  `o.Namespace` is the default.

File: pkg/oc/cli/debug/debug.go (hunk 2 above)
Comment (smarterclayton) [reply pkg/oc/cli/debug/debug.go]: restructured it

File: pkg/oc/cli/debug/debug.go (hunk 3 above)
Comment (smarterclayton) [reply pkg/oc/cli/debug/debug.go]: We need to account for the resources we actually use, but without adding `--resources` to allow an admin to override it's probably too early.  The tradeoff is that we might get evicted if the node is under pressure, but it's slightly better for the workloads on the node if the admin debug pod is what gets evicted, not the workload.  

We will have to have some future way to ensure we don't get evicted, but I don't want to design it yet until we know how frequent this issue is.  As a workaround an admin can always `-o yaml` and edit the pod, then start and attach.

File: pkg/oc/cli/debug/debug.go (hunk 5 above)
Comment (smarterclayton) [reply pkg/oc/cli/debug/debug.go]: It doesn't only apply to node debugging.  It is used in transform podForDebug.

Primary use case is you want to debug the volume mounts of a given pod but with a different image.

**Comment by smarterclayton** [comment]
I don't know what you mean by "I can't run crictl" in this pod?

**Comment by sjenning** [comment]
> I can't run crictl in this pod, so i can't look at failing static pod logs

don't have a cluster atm to test on but can you just do `chroot /host /bin/bash` and get `crictl` on the PATH for free?

**Comment by deads2k** [comment]
> don't have a cluster atm to test on but can you just do `chroot /host /bin/bash` and get `crictl` on the PATH for free?

Yes, this works.  Thanks.

**Comment by smarterclayton** [comment]
Yes you can.  I will test that.

Updated with comments addressed.

File: pkg/oc/cli/debug/debug.go (hunk 4 above)
Comment (smarterclayton) [reply pkg/oc/cli/debug/debug.go]: Simplified

**Comment by smarterclayton** [comment]
Yes, `chroot /host` inside the debug pod worked for me

**Comment by smarterclayton** [comment]
Tag or comment

**Comment by smarterclayton** [comment]
Clearing the queue, this has passed multiple times and has no impact on e2e

**Comment by akostadinov** [comment]
If I run `oc --config=/tmp/v4.kubeconfig debug node/ip-10-0-177-220.us-east-2.compute.internal` on two terminals, then the debug pod is killed. Can we have more than one debug pod per node, or same pod handle multiple connections?

**Comment by jeremyeder** [comment]
Thanks, this works now after https://github.com/openshift/origin/pull/22086 was merged. It is AWESOME that you're using rhel-tools for this!

I know that ```oc debug node/NODE_NAME --image=local.registry/busybox``` works.

Should we worry about the scenario when I can't reach an external registry?  Should we automatically populate the OCP registry with a rhel-tools image, for example?
